src/expr/sygus_grammar.cpp

```diff
@@ -0,0 +1,175 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Abdalrhman Mohamed
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of class for constructing SyGuS Grammars.
+ */
+
+#include "expr/sygus_grammar.h"
+
+#include <sstream>
+
+#include "printer/printer.h"
+#include "printer/smt2/smt2_printer.h"
+
+namespace cvc5::internal {
+
+SygusGrammar::SygusGrammar(const std::vector<Node>& sygusVars,
+                           const std::vector<Node>& ntSyms)
+    : d_sygusVars(sygusVars), d_ntSyms(ntSyms)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  for (const Node& ntSym : ntSyms)
+  {
+    d_sdts.emplace(ntSym, SygusDatatype(ntSym.getName()));
+    d_ntsToUnres.emplace(ntSym, nm->mkUnresolvedDatatypeSort(ntSym.getName()));
+  }
+}
+
+bool isId(const Node& n)
+{
+  return n.getKind() == kind::LAMBDA && n[0].getNumChildren() == 1
+         && n[0][0] == n[1];
+}
+
+void SygusGrammar::addRule(const Node& ntSym, const Node& rule)
```

Comment:
Add an assertion that `ntSym` is in d_sdts.

(This is checked at the API but I foresee new internal usages of this class, in which case we should be sure this is called correctly).